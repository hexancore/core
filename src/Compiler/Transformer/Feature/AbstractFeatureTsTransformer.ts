import ts from 'typescript';
import type { FeatureMap, FeatureMeta } from "../../../Util/Feature/Meta/FeatureMeta";
import { ModuleClassTsTransformer } from '../ModuleClassTsTransformer';
import type { TsImportHelper } from '../TsImportHelper';
import type { FeatureTransformContext } from './FeatureTransformContext';
import type { FeatureSourcePath } from "../../../Util/Feature/FeatureModuleDiscoverer";

export interface FeatureTsTransformerHelpers {
  importHelper: TsImportHelper;
  moduleClassTransformer: ModuleClassTsTransformer;
  features: FeatureMap;
}

export abstract class AbstractFeatureTsTransformer {
  protected importHelper: TsImportHelper;
  protected moduleClassTransformer: ModuleClassTsTransformer;
  protected features: FeatureMap;

  public constructor(helpers: FeatureTsTransformerHelpers) {
    this.importHelper = helpers.importHelper;
    this.moduleClassTransformer = helpers.moduleClassTransformer;
    this.features = helpers.features;
  }

  public abstract transform(source: ts.SourceFile, context: FeatureTransformContext): ts.SourceFile;

  public abstract supports(featureSourcePath: FeatureSourcePath, feature: FeatureMeta): boolean;

}