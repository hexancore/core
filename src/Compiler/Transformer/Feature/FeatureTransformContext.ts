import type ts from "typescript";
import type { FeatureMeta } from "../../../Util/Feature/Meta/FeatureMeta";

export interface FeatureTransformContext {
  feature: FeatureMeta;
  sourceFilePathWithoutRoot: string;
  tsContext: ts.TransformationContext;
  diagnostics: ts.Diagnostic[];
}