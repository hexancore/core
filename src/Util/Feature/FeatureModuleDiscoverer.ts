import { type PathsOutput } from "fdir";
import path from "node:path";
import { FsHelper } from "../Filesystem/FsHelper";
import { PerformanceHelper } from "../Performance/PerformanceHelper";
import { FeatureApplicationDiscoverer } from "./FeatureApplicationDiscoverer";
import { FeatureDiscovererHelper } from "./FeatureDiscovererHelper";
import { FeatureDomainDiscoverer } from "./FeatureDomainDiscoverer";
import { FeatureInfrastructureDiscoverer } from "./FeatureInfrastructureDiscoverer";
import { FeatureMeta } from "./Meta/FeatureMeta";

export const CORE_FEATURE_NAME = "Core";

export class FeatureModuleDiscoverer {

  private applicationDiscoverer: FeatureApplicationDiscoverer;
  private domainDiscoverer: FeatureDomainDiscoverer;
  private infrastructureDiscoverer: FeatureInfrastructureDiscoverer;

  public constructor(private sourceRoot: string) {
    this.sourceRoot = FsHelper.normalizePathSep(this.sourceRoot);

    this.applicationDiscoverer = new FeatureApplicationDiscoverer();
    this.domainDiscoverer = new FeatureDomainDiscoverer();
    this.infrastructureDiscoverer = new FeatureInfrastructureDiscoverer();
  }

  public static extractFeatureNameFromPath(sourceRoot: string, sourcePath: string): string | null {
    if (!sourcePath.startsWith(sourceRoot)) {
      return null;
    }

    const withoutSourceRootParts = sourcePath.substring(sourceRoot.length + 1).split("/", 2);
    if (withoutSourceRootParts.length < 2) {
      return null;
    }

    const featureName = withoutSourceRootParts[0];
    return featureName === CORE_FEATURE_NAME ? null : featureName;
  }

  public async discoverAll(): Promise<Map<string, FeatureMeta>> {
    const featureDirs: PathsOutput = await FeatureDiscovererHelper.crawlDirOnly(this.sourceRoot);
    const features = new Map<string, FeatureMeta>();
    for (const featureDir of featureDirs) {
      const featureName = path.basename(featureDir);
      if (featureName === CORE_FEATURE_NAME) {
        continue;
      }

      const discovery = await this.discover(featureName, featureDir);
      features.set(featureName, discovery);
    }

    return features;
  }

  private async discover(featureName: string, featureDir: string): Promise<FeatureMeta> {
    featureDir = featureDir.endsWith("/") ? featureDir.substring(0, featureDir.length - 1) : featureDir;

    return PerformanceHelper.measureFunction("hc.core.feature_module_discovery" + featureName, async () => {
      return new FeatureMeta(
        featureName,
        await this.applicationDiscoverer.discover(featureDir),
        await this.domainDiscoverer.discover(featureDir),
        await this.infrastructureDiscoverer.discover(featureDir, featureName)
      );
    });
  }
}