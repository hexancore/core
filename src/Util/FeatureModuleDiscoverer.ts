import { fdir, type PathsOutput } from "fdir";
import { pathExists } from 'fs-extra';
import path from "node:path";
import { PerformanceHelper } from "./Performance/PerformanceHelper";
import { hash } from "node:crypto";
import { readFile } from 'node:fs/promises';

export interface ClassDiscovery {
  className: string;
  path: string;
}

export interface FeatureApplicationServiceInfo {
  className: string;
  /**
   * Name of context where service belongs.
   */
  context: string;
  /**
   * Indicate is marked with `@Injectable`
   */
  isInjectable: boolean;

  /**
   * Relative to feature root.
   */
  path: string;
}

export interface FeatureApplicationMessageInfo {
  context: string;
  className: string,
  handlerClassName: string,
  name: string;
  path: string;
}

export interface FeatureModuleDiscovery {
  application: {
    commands: FeatureApplicationMessageInfo[];
    queries: FeatureApplicationMessageInfo[];
    events: FeatureApplicationMessageInfo[];
    services: FeatureApplicationServiceInfo[];
  };

  infrastructure: {
    module: ClassDiscovery;
  };

  cacheKey: string;
}

export class FeatureModuleDiscoverer {
  public constructor(private sourceRoot: string) {
    this.sourceRoot.replaceAll("\\", "/");
  }

  public async discoverAll(): Promise<Map<string, FeatureModuleDiscovery>> {
    const featureDirs: PathsOutput = await (new fdir())
      .onlyDirs()
      .withMaxDepth(1)
      .withPathSeparator("/")
      .crawl(this.sourceRoot)
      .withPromise();

    featureDirs.shift();
    const features = new Map<string, FeatureModuleDiscovery>();
    for (const featureDir of featureDirs) {
      const discovery = await FeatureModuleDiscoverer.discover(featureDir);
      features.set(path.basename(featureDir), discovery);
    }

    return features;
  }

  public static async discover(featureDir: string): Promise<FeatureModuleDiscovery> {
    const featureName = path.basename(featureDir);
    featureDir = featureDir.replaceAll("\\", "/");
    featureDir = featureDir.endsWith("/") ? featureDir.substring(0, featureDir.length - 1) : featureDir;
    return PerformanceHelper.measureFunction("hc.core.feature_module_discovery" + featureName, async () => {
      const contextDirs: PathsOutput = await (new fdir())
        .onlyDirs()
        .withMaxDepth(1)
        .withPathSeparator("/")
        .crawl(path.join(featureDir, "Application"))
        .withPromise();

      const commands: FeatureApplicationMessageInfo[] = [];
      const queries: FeatureApplicationMessageInfo[] = [];
      const events: FeatureApplicationMessageInfo[] = [];
      const services: FeatureApplicationServiceInfo[] = [];

      for (const contextDir of contextDirs) {
        const contextName = path.basename(contextDir);
        commands.push(...(await this.discoverMessageHandlers(contextName, contextDir, featureDir, 'Command')));
        queries.push(...(await this.discoverMessageHandlers(contextName, contextDir, featureDir, 'Query')));
        events.push(...(await this.discoverMessageHandlers(contextName, contextDir, featureDir, 'Event')));
        services.push(...(await this.discoverServices(contextName, contextDir, featureDir)));
      }

      const discovery = {
        application: {
          commands,
          queries,
          events,
          services,
        },
        infrastructure: {
          module: this.discoverInfrastructureModule(featureName)
        }
      };

      this.calcCacheHashKey(discovery);
      return discovery as FeatureModuleDiscovery;
    });
  }

  public static extractFeatureNameFromPath(sourceRoot: string, sourcePath: string): string | null {
    if (sourcePath.startsWith(sourceRoot) && sourcePath.endsWith("Module.ts")) {
      const withoutSourceRoot = sourcePath.substring(sourceRoot.length + 1);
      const withoutSourceRootSplit = withoutSourceRoot.split("/");
      return withoutSourceRootSplit.length === 2 ? withoutSourceRootSplit[0] : null;
    }

    return null;
  }

  private static async discoverMessageHandlers(context: string, contextDir: string, featureDir: string, messageType: 'Command' | 'Query' | 'Event'): Promise<FeatureApplicationMessageInfo[]> {
    const dir = contextDir + `${messageType}`;
    if (!(await pathExists(dir))) {
      return [];
    }

    const dirIt = (new fdir()).withPathSeparator("/").onlyDirs().withMaxDepth(1).withFullPaths();
    const messagePaths = await dirIt.crawl(dir).withPromise();
    messagePaths.shift();
    const messages: FeatureApplicationMessageInfo[] = [];
    for (const messagePath of messagePaths) {
      const name = path.basename(messagePath);
      messages.push({
        context: context,
        name: path.basename(messagePath),
        className: `${context}${name}${messageType}`,
        handlerClassName: `${context}${name}${messageType}Handler`,
        path: messagePath.substring(featureDir.length + 1, messagePath.length - 1),

      });
    }

    return messages;
  }

  private static async discoverServices(context: string, contextDir: string, featureDir: string): Promise<FeatureApplicationServiceInfo[]> {
    const dir = contextDir + `Service`;
    if (!(await pathExists(dir))) {
      return [];
    }

    const dirIt = (new fdir()).withPathSeparator("/").withFullPaths();
    dirIt.filter((path) => !path.endsWith("index.ts"));

    const filePaths = await dirIt.crawl(dir).withPromise();
    const classes: FeatureApplicationServiceInfo[] = [];
    for (const filePath of filePaths) {
      const isInjectable = (await readFile(filePath)).indexOf("@Injectable(", 0, 'utf8') !== -1;
      classes.push({
        className: path.basename(filePath, ".ts"),
        context: context,
        isInjectable,
        path: filePath.substring(featureDir.length + 1, filePath.length - 3),
      });
    }

    return classes;
  }

  private static discoverInfrastructureModule(featureName: string): ClassDiscovery {
    const className = `${featureName}InfrastructureModule`;
    const filePath = `Infrastructure/${className}`;
    return { className, path: filePath };
  }

  private static calcCacheHashKey(discovery: Omit<FeatureModuleDiscovery, 'cacheKey'>) {
    let hashSource = "";
    for (const c of discovery.application.commands) {
      hashSource += c.context + c.name;
    }
    for (const c of discovery.application.queries) {
      hashSource += c.context + c.name;
    }
    for (const c of discovery.application.events) {
      hashSource += c.context + c.name;
    }
    for (const c of discovery.application.services) {
      hashSource += c.path;
    }

    discovery["cacheKey"] = hash("sha1", hashSource, "hex");
  }

}