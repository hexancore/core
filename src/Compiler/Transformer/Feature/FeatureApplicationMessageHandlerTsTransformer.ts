import { FeatureApplicationCommandMeta, type FeatureApplicationQueryMeta } from '../../../Util/Feature/Meta';
import ts from "typescript";
import { type FeatureMeta } from "../../../Util/Feature/Meta";
import { AbstractFeatureTsTransformer } from "./AbstractFeatureTsTransformer";
import type { FeatureTransformContext } from "./FeatureTransformContext";
import type { FeatureSourcePath } from "../../../Util/Feature/FeatureModuleDiscoverer";
import { ImportDeclarationWrapper } from "../Helper/ImportDeclarationWrapper";

interface VisitContext extends FeatureTransformContext {
  meta: FeatureApplicationCommandMeta | FeatureApplicationQueryMeta;
  handlerDecoratorName: string;
  cqrsImport: ImportDeclarationWrapper;
}

export class FeatureApplicationMessageHandlerTsTransformer extends AbstractFeatureTsTransformer {

  public supports(featureSourcePath: FeatureSourcePath, feature: FeatureMeta): boolean {
    const meta = feature.applicationCqrsHandlerMap.get(featureSourcePath.localSourcePath);
    return !!meta;
  }

  public transform(source: ts.SourceFile, context: FeatureTransformContext): ts.SourceFile {
    const messageMeta = context.feature.applicationCqrsHandlerMap.get(context.featureSourcePath.localSourcePath)!;
    const handlerDecoratorName = messageMeta instanceof FeatureApplicationCommandMeta ? "CommandHandler" : "QueryHandler";
    const cqrsImport = ImportDeclarationWrapper.create(handlerDecoratorName, "@nestjs/cqrs", context.tsContext);
    const visitContext: VisitContext = {
      ...context,
      meta: messageMeta,
      handlerDecoratorName: handlerDecoratorName,
      cqrsImport
    };

    const transformed = ts.factory.updateSourceFile(source, [
      cqrsImport.declaration,
      ...ts.visitNodes(
        source.statements,
        this.visitSourceStatement.bind(this, visitContext),
        (node): node is ts.Statement => ts.isStatement(node)
      )
    ]);

    return transformed;
  }

  private visitSourceStatement(visitContext: VisitContext, node: ts.Statement) {
    if (ts.isClassDeclaration(node)) {
      const handlerDecorator = ts.factory.createDecorator(ts.factory.createCallExpression(
        visitContext.cqrsImport.get(visitContext.handlerDecoratorName),
        undefined,
        [ts.factory.createIdentifier(visitContext.meta.className)]
      ));
      return ts.factory.updateClassDeclaration(
        node,
        [handlerDecorator, ...node.modifiers ?? []],
        node.name,
        node.typeParameters,
        node.heritageClauses,
        node.members
      );
    }

    return node;
  }
}