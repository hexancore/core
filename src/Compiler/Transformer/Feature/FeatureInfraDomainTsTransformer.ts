import ts from 'typescript';
import type { FeatureMeta } from "../../../Util/Feature/Meta/FeatureMeta";
import { AbstractFeatureTsTransformer } from './AbstractFeatureTsTransformer';
import { TsTransfromerHelper } from '../TsTransformerHelper';
import type { AddImportTransformDef } from '../ModuleClassTsTransformer';
import type { FeatureTransformContext } from './FeatureTransformContext';

export class FeatureInfraDomainModuleTsTransformer extends AbstractFeatureTsTransformer {

  public transform(source: ts.SourceFile, context: FeatureTransformContext): ts.SourceFile {
    const feature = context.feature;
    const domainErrorsClassName = feature.name + "DomainErrors";

    const imports: AddImportTransformDef[] = [
      { name: "DomainInfraModuleHelper", importModuleSpecifier: '@hexancore/core' },
      { name: domainErrorsClassName, importModuleSpecifier: '../../Domain' }
    ];

    const repos: string[] = [];

    for (const r of feature.domain.aggregateRoots) {
      repos.push(r.infraRepositoryName);
      imports.push(
        { name: r.name, importModuleSpecifier: `../../Domain/${r.name}/${r.name}` },
        { name: r.infraRepositoryName, importModuleSpecifier: `./${r.name}/${r.infraRepositoryName}` }
      );

      for (const e of r.entities) {
        repos.push(e.infraRepositoryName);
        imports.push({ name: e.infraRepositoryName, importModuleSpecifier: `./${r.name}/${e.infraRepositoryName}` });
      }
    }

    return this.moduleClassTransformer.transform({
      imports,
      extraStatementProvider(importedIdentifierMapper) {
        const classIdentifier = importedIdentifierMapper("DomainInfraModuleHelper");
        const methodIdentifier = ts.factory.createIdentifier("createMeta");

        const optionsObject = ts.factory.createObjectLiteralExpression([
          ts.factory.createPropertyAssignment("featureName", ts.factory.createStringLiteral(feature.name)),
          ts.factory.createPropertyAssignment("aggregateRootCtrs", ts.factory.createArrayLiteralExpression(feature.domain.aggregateRoots.map((r) => ts.factory.createIdentifier(r.name)))),
          ts.factory.createPropertyAssignment("domainErrors", ts.factory.createIdentifier(domainErrorsClassName)),
        ]);

        const createMeta = TsTransfromerHelper.createConstStatement("HcDomainInfraModuleMetaExtra", ts.factory.createCallExpression(
          ts.factory.createPropertyAccessExpression(classIdentifier, methodIdentifier),
          undefined,
          [optionsObject]
        ));

        return [
          TsTransfromerHelper.createConstStatement("HcDomainInfraAggrgateRootRepositories",
            ts.factory.createArrayLiteralExpression(repos.map((r) => ts.factory.createIdentifier(r)))
          ),
          createMeta
        ];
      },

      extraMetaProvider() {
        return ts.factory.createIdentifier("HcDomainInfraModuleMetaExtra");
      },
      source,
      context: context.tsContext,
    });
  }

  public supports(sourcefilePath: string, feature: FeatureMeta): boolean {
    return sourcefilePath.endsWith("DomainInfraModule.ts");
  }
}