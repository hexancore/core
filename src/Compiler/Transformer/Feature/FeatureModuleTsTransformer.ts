import ts from 'typescript';
import type {FeatureMeta } from "../../../Util/Feature/Meta/FeatureMeta";
import { AbstractFeatureTsTransformer } from "./AbstractFeatureTsTransformer";
import type { ProviderModuleMetaTransformDef } from '../ModuleClassTsTransformer';
import type { FeatureTransformContext } from './FeatureTransformContext';
import type { FeatureApplicationMessageMeta } from '../../../Util/Feature/Meta';
import type { FeatureSourcePath } from "../../../Util/Feature/FeatureModuleDiscoverer";

/**
 * Adding automatic injection of message handlers, services, infra module to `[Feature]Module` source.
 * Less write, more fun !
 */
export class FeatureModuleTsTransformer extends AbstractFeatureTsTransformer {

  public supports(featureSourcePath: FeatureSourcePath, feature: FeatureMeta): boolean {
    return featureSourcePath.localSourcePath.endsWith(feature.name + "Module.ts");
  }

  public transform(source: ts.SourceFile, context: FeatureTransformContext): ts.SourceFile {
    const feature = context.feature;

    const messageHandlersProviders: ProviderModuleMetaTransformDef[] = [];
    messageHandlersProviders.push(...this.createMessageHandlerProviders(feature.application.commands));
    messageHandlersProviders.push(...this.createMessageHandlerProviders(feature.application.queries));

    return this.moduleClassTransformer.transform({
      imports: [],
      meta: {
        imports: [],
        providers: [...messageHandlersProviders, ...this.createServiceProviders(feature)],
      },
      source,
      context: context.tsContext,
    });
  }

  private createMessageHandlerProviders(messages: FeatureApplicationMessageMeta[]): ProviderModuleMetaTransformDef[] {
    const providers: ProviderModuleMetaTransformDef[] = [];
    for (const m of messages) {
      const importPath = `./${m.path}/${m.handlerClass}`;
      providers.push({
        addToExports: false,
        name: m.handlerClass,
        importFrom: importPath
      });
    }

    return providers;
  }

  private createServiceProviders(feature: FeatureMeta): ProviderModuleMetaTransformDef[] {
    const providers: ProviderModuleMetaTransformDef[] = [];
    for (const s of feature.application.services) {
      if (!s.isInjectable) {
        continue;
      }

      providers.push({
        addToExports: false,
        name: s.name,
        importFrom: `./${s.path}`
      });
    }

    return providers;
  }
}