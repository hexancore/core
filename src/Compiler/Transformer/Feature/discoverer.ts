import { FeatureModuleDiscoverer } from "../../../Util/Feature/FeatureModuleDiscoverer";
// node ./lib/Compiler/Transformer/Feature/discoverer.js ./test/helper/libs/test-lib/src
async function run() {
  if (process.argv.length < 3) {
    throw new Error("Wrong execute, not given sourceRoot argument");
  }
  const sourceRoot = process.argv[2];
  const discoverer = new FeatureModuleDiscoverer(sourceRoot);
  const features = await discoverer.discoverAll();
  console.log(JSON.stringify(Array.from(features.entries())));
}

run();