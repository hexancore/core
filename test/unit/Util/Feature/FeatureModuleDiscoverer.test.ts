/**
 * @group unit
 */

import { FeatureModuleDiscoverer } from "@/Util/Feature/FeatureModuleDiscoverer";
import { LIBS_DIRNAME } from "@test/libs";

describe(FeatureModuleDiscoverer.constructor.name, () => {
  test('discoverAll', async () => {
    const discoverer = new FeatureModuleDiscoverer(LIBS_DIRNAME + "/test-lib/src");

    const current = await discoverer.discoverAll();

    expect(current).toMatchSnapshot();
  });
});
