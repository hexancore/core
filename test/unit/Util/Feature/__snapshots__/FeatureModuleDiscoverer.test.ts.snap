// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Function discoverAll 1`] = `
"- name: Book
  cacheKey: b44e0f68146acb946dab46987c547d2664ce6088
  application:
    commands:
      - name: Create
        context: Book
    queries:
      - name: GetById
        context: Book
    dtos:
      - name: ArrayItemsTestDto
        context: Book
        path: Application/Book/Dto/ArrayItemsTestDto
      - name: BookDto
        context: Book
        path: Application/Book/Dto/BookDto
      - name: FloatTestDto
        context: Book
        path: Application/Book/Dto/FloatTestDto
      - name: IntTestDto
        context: Book
        path: Application/Book/Dto/IntTestDto
      - name: StringTestDto
        context: Book
        path: Application/Book/Dto/StringTestDto
      - name: TestTransformDto
        context: Book
        path: Application/Book/Dto/TestTransformDto
      - name: UIntTestDto
        context: Book
        path: Application/Book/Dto/UIntTestDto
    services:
      - name: BookInfraPort
        context: Book
        path: Application/Book/Service/BookInfraPort
        isInjectable: false
      - name: BookService
        context: Book
        path: Application/Book/Service/BookService
        isInjectable: true
  domain:
    aggregateRoots:
      - name: Book
        entities:
          - name: BookCopy
            aggregateRootName: Book
        valueObjects:
          - name: BookCopyId
            context: Book
            shared: true
          - name: BookId
            context: Book
            shared: true
    valueObjects:
      - name: FeatureSharedValueObject
        context: ''
        shared: true
  infrastructure:
    module:
      name: BookInfraModule
      path: Infrastructure/BookInfraModule
      filePath: Infrastructure/BookInfraModule.ts
"
`;

exports[`Function hObjectMap 1`] = `
"- Application/Book/Command/Create/BookCreateCommand.ts:
    name: Create
    context: Book
    className: BookCreateCommand
    handlerClass: BookCreateCommandHandler
    path: Application/Book/Command/Create
  Application/Book/Query/GetById/BookGetByIdQuery.ts:
    name: GetById
    context: Book
    className: BookGetByIdQuery
    handlerClass: BookGetByIdQueryHandler
    path: Application/Book/Query/GetById
  Application/Book/Dto/ArrayItemsTestDto.ts:
    name: ArrayItemsTestDto
    context: Book
    path: Application/Book/Dto/ArrayItemsTestDto
    layer: application
  Application/Book/Dto/BookDto.ts:
    name: BookDto
    context: Book
    path: Application/Book/Dto/BookDto
    layer: application
  Application/Book/Dto/FloatTestDto.ts:
    name: FloatTestDto
    context: Book
    path: Application/Book/Dto/FloatTestDto
    layer: application
  Application/Book/Dto/IntTestDto.ts:
    name: IntTestDto
    context: Book
    path: Application/Book/Dto/IntTestDto
    layer: application
  Application/Book/Dto/StringTestDto.ts:
    name: StringTestDto
    context: Book
    path: Application/Book/Dto/StringTestDto
    layer: application
  Application/Book/Dto/TestTransformDto.ts:
    name: TestTransformDto
    context: Book
    path: Application/Book/Dto/TestTransformDto
    layer: application
  Application/Book/Dto/UIntTestDto.ts:
    name: UIntTestDto
    context: Book
    path: Application/Book/Dto/UIntTestDto
    layer: application
  Domain/Book/Shared/ValueObject/BookCopyId.ts:
    name: BookCopyId
    context: Book
    shared: true
  Domain/Book/Shared/ValueObject/BookId.ts:
    name: BookId
    context: Book
    shared: true
  Domain/Shared/ValueObject/FeatureSharedValueObject.ts:
    name: FeatureSharedValueObject
    context: ''
    shared: true
"
`;
